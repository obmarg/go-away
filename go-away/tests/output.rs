#![allow(dead_code)]

use insta::assert_snapshot;

use go_away::{TypeMetadata, TypeRegistry};

#[derive(TypeMetadata)]
struct MyData {
    field_one: String,
    nested: Nested,
}

#[test]
fn test_struct_output() {
    let mut registry = TypeRegistry::new();
    MyData::metadata(&mut registry);

    assert_snapshot!(go_away::registry_to_output(registry));
}

#[derive(TypeMetadata)]
#[serde(tag = "type", content = "data")]
enum NewTypeEnum {
    OptionOne(One),
    OptionTwo(Two),
}

#[derive(TypeMetadata)]
struct Nested {
    #[serde(rename = "some_other_name")]
    a_string: String,
    an_int: i64,
    fulfilment_type: FulfilmentType,
}

#[derive(TypeMetadata)]
enum FulfilmentType {
    Delivery,
    Collection,
}

#[derive(TypeMetadata)]
struct One {
    x: f32,
}

#[derive(TypeMetadata)]
struct Two {
    y: bool,
}

#[test]
fn test_newtype_enum() {
    let mut registry = TypeRegistry::new();
    NewTypeEnum::metadata(&mut registry);

    assert_snapshot!(go_away::registry_to_output(registry));
}

#[derive(TypeMetadata)]
#[serde(tag = "type", content = "data")]
enum StructEnum {
    OptionOne { x: String, y: i32 },
    OptionTwo { foo: String, bar: Nested },
}

#[test]
fn test_struct_enum() {
    let mut registry = TypeRegistry::new();
    StructEnum::metadata(&mut registry);

    assert_snapshot!(go_away::registry_to_output(registry));
}

#[derive(TypeMetadata)]
#[serde(tag = "type")]
enum InternallyTaggedTupleEnum {
    One(One),
    Two(Two),
}

#[test]
fn test_internally_tagged_tuple_enum() {
    let mut registry = TypeRegistry::new();
    InternallyTaggedTupleEnum::metadata(&mut registry);

    assert_snapshot!(go_away::registry_to_output(registry));
}
