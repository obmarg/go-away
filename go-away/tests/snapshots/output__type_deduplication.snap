---
source: go-away/tests/output.rs
expression: "go_away::registry_to_output(registry)"

---
type MyData struct {
	FieldOne string `json:"field_one"`
	Nested Nested `json:"nested"`
}
type OptionTwo struct {
	Foo string `json:"foo"`
	Bar Nested `json:"bar"`
}
type Nested struct {
	AString string `json:"some_other_name"`
	AnInt int `json:"an_int"`
	FulfilmentType FulfilmentType `json:"fulfilment_type"`
}
type OptionOne struct {
	X string `json:"x"`
	Y int `json:"y"`
}
type StructEnum struct {
	OptionOne *OptionOne
	OptionTwo *OptionTwo
}

func (self *StructEnum) MarshalJSON() ([]byte, error) {
	if self.OptionOne != nil {
		var output map[string]interface{}
		output["type"] = "OptionOne"
		output["data"] = self.OptionOne
		return json.Marshal(output)
	} else 	if self.OptionTwo != nil {
		var output map[string]interface{}
		output["type"] = "OptionTwo"
		output["data"] = self.OptionTwo
		return json.Marshal(output)
	} else 	{
		return json.Marshal(nil)
	}
}
func (self *StructEnum) UnmarshalJSON(data []byte) error {
	temp := struct{
		Tag string `json:"type"`
	}{}
	if err := json.Unmarshal(data, &temp); err != nil {
		return err
	}
	if temp.Tag == "OptionOne" {
		rv := struct{
			Data OptionOne `json:"data"`
		}{}
		if err := json.Unmarshal(data, &rv); err != nil {
			return err
		}
		self.OptionOne = &rv.Data
		self.OptionTwo = nil
	} else if temp.Tag == "OptionTwo" {
		rv := struct{
			Data OptionTwo `json:"data"`
		}{}
		if err := json.Unmarshal(data, &rv); err != nil {
			return err
		}
		self.OptionTwo = &rv.Data
		self.OptionOne = nil
	} else {
		return errors.New("Unknown type tag")
	}
	return nil
}
type FulfilmentType string

const (
	Delivery FulfilmentType = "Delivery"
	Collection FulfilmentType = "Collection"
)

